app/services/add_attachment_service.rb

```diff
@@ -0,0 +1,63 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddAttachmentService
+  attr_reader :container, :author
+
+  def initialize(container, author:)
+    @container = container
+    @author = author
+  end
+
+  ##
+  # Adds and saves the uploaded file as attachment of the given container.
+  # In case the container supports it, a journal will be written.
+  #
+  # An ActiveRecord::RecordInvalid error is raised if any record can't be saved.
+  def add_attachment(uploaded_file:, description:)
+    attachment = Attachment.new(file: uploaded_file,
+                                container: container,
+                                description: description,
+                                author: author)
+    save attachment
```

Comment:
This is only my humble opinion, but: 
Why do you let the internal process escalate into outside the service? 

If you use the AR semantics of `save`, you should catch the exception inside save and return false. The calling attachment endpoint could then look like before.

``` ruby
service = AddAttachmentService.new(
  @work_package,
  author: current_user)
  uploaded_file: uploaded_file,
  description: metadata.description
)

unless service.call
  raise ::API::Errors::ErrorBase.create_and_merge_errors(attachment.errors)
end
```

Now, you get the attachment from the service, even though you don't need it and the API endpoint needs to know more about the service's internals (which exceptions it raises) than necessary.